hail/src/main/scala/is/hail/expr/ir/Emit.scala

```diff
@@ -857,6 +857,199 @@ class Emit[C](
+              val leftDataAddress = lPType.data.load(leftPVal.tcode[Long])
+              val rightDataAddress = rPType.data.load(rightPVal.tcode[Long])
+
+              val answerPArrayAddress = mb.genFieldThisRef[Long]()
+              val M = lShape(lPType.nDims - 2)
+              val N = rShape(rPType.nDims - 1)
+              val K = lShape(lPType.nDims - 1)
+
+              val LDA = M
+              val LDB = K
+              val LDC = M
+
+              val multiplyViaDGEMM = Code(
+                ((M cne 0L) && (N cne 0L) && (K cne 0L)).mux(Code(
+                  answerPArrayAddress := outputPType.data.pType.allocate(region.code, (M * N).toI),
+                  outputPType.data.pType.stagedInitialize(answerPArrayAddress, (M * N).toI),
+                  lPType.elementType match {
+                    case PFloat32(_) =>
+                      Code.invokeScalaObject13[String, String, Int, Int, Int, Float, Long, Int, Long, Int, Float, Long, Int, Unit](BLAS.getClass, method = "sgemm",
+                        "N",
+                        "N",
+                        M.toI,
+                        N.toI,
+                        K.toI,
+                        1.0f,
+                        lPType.data.pType.firstElementOffset(leftDataAddress),
+                        LDA.toI,
+                        rPType.data.pType.firstElementOffset(rightDataAddress),
+                        LDB.toI,
+                        0.0f,
+                        outputPType.data.pType.firstElementOffset(answerPArrayAddress, (M * N).toI),
+                        LDC.toI
+                      )
+                    case PFloat64(_) =>
+                      Code.invokeScalaObject13[String, String, Int, Int, Int, Double, Long, Int, Long, Int, Double, Long, Int, Unit](BLAS.getClass, method = "dgemm",
+                        "N",
+                        "N",
+                        M.toI,
+                        N.toI,
+                        K.toI,
+                        1.0,
+                        lPType.data.pType.firstElementOffset(leftDataAddress),
+                        LDA.toI,
+                        rPType.data.pType.firstElementOffset(rightDataAddress),
+                        LDB.toI,
+                        0.0,
+                        outputPType.data.pType.firstElementOffset(answerPArrayAddress, (M * N).toI),
+                        LDC.toI
+                      )
+                  }),
+                  answerPArrayAddress := outputPType.data.pType.zeroes(mb,
+                    region.code, (M * N).toI)
+                ),
+                outputPType.construct(
+                  outputPType.makeShapeBuilder(IndexedSeq(M, N)),
+                  outputPType.makeColumnMajorStridesBuilder(IndexedSeq(M, N), mb),
+                  answerPArrayAddress,
+                  mb,
+                  region.code))
+
+              EmitCode(Code._empty, false, PCode(pt, multiplyViaDGEMM)).toI(cb)
```

Comment:
I'm not super excited by the idea of mixing the two paradigms here. If you need to use the NDArrayEmitter here, it might be a good opportunity to introduce a mostly imperative version before you port this since it sounds like that's the direction you're heading.

Reply:
When you say two paradigms, do you mean that I am making an `EmitCode` still? Or just generally that I'm mostly still using giant `Code` blocks? I can clean up this top part you highlighted some more and not construct an `EmitCode`, but the below emitter parts would require some emitter updates. 

As for direction I'm heading, mostly I was trying to get all the NDArray stuff a little cleaner and into `emitI`, at which point I was planning to make a follow up PR that pulled all the NDArray emission stuff into something called `emitNDArray` so that I can make sure any final NDArray results that get emitted come back column major, even if it's possible for intermediate NDArrays not to be. I wanted to do that so that we can remove `strides` from `TNDArray`, where it certainly doesn't belong. I was also going to go off and try to create an imperative version of `NDArrayEmitter`.

Reply:
Yes, both. 

My suggestion would be to get the imperative NDArray emitter in before/simultaneous with this change. Afaict, that's not blocked on this change (although feel free to push back if it turns out I'm wrong!) but will make future lifts smoother and more complete (including this one!).